File: java/130_Surrounded_Regions.java
package java;

class Solution {
    int n = 0;
    int m = 0;
    public void solve(char[][] board) {
        n = board.length;
        m = board[0].length;
        if (n == 0){
            return ;
        }
        for (int i = 0 ; i < n ; i++){
            dfs(board, i, 0);
            dfs(board, i, m-1);
        }
        for (int i = 0 ; i < m ; i++){
            dfs(board, 0, i);
            dfs(board, n-1, i);
        }
        for (int i = 0 ; i < n ; i++){
            for(int j = 0 ; j < m ; j++){
                if(board[i][j] == 'A'){
                    board[i][j] = 'O';
                }
                else if(board[i][j] == 'O'){
                    board[i][j] = 'X';
                }
            }
        }
    }

    public void dfs(char[][] board, int x, int y){
        if(x < 0 || x >= n || y < 0 || y >= m || board[x][y] != 'O'){
            return;
        }
        board[x][y] = 'A';
        int[] dx = {0,1,0,-1};
        int[] dy = {1,0,-1,0};
        for(int i = 0 ; i < 4 ;i++){
            dfs(board, x+dx[i], y+dy[i]);
        }
    }
}

File: python/130_Surrounded_Regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        if not board:
            return
        
        row , col = len(board) , len(board[0])
        
        def dfs(x,y): 
            # mark all egde 'O's as 'A'
            if(x < 0 or x >= row or y < 0 or y >= col):
                return
            if board[x][y] != 'O':
                return
            
            dx = [0,1,0,-1]
            dy = [1,0,-1,0]
            board[x][y] = 'A'
            
            for i in range(4):
                dfs(x+dx[i],y+dy[i])
        
        def bfs(x,y):
            from collections import deque
            queue = deque()
            queue.append((x,y))
            while queue:
                # mark all egde 'O's as 'A'
                (x,y) = queue.pop()
                if(x < 0 or x >= row or y < 0 or y >= col):
                    continue
                if board[x][y] != 'O':
                    continue

                dx = [0,1,0,-1]
                dy = [1,0,-1,0]
                board[x][y] = 'A'

                for i in range(4):
                    queue.append((x+dx[i],y+dy[i]))
        
        for i in range(row):
            bfs(i,0)
            bfs(i,col-1)
        
        for i in range(col):
            bfs(0,i)
            bfs(row-1,i)
        
        for x in range(row):
            for y in range(col):
                if board[x][y] == 'A':
                   board[x][y] = 'O'
                elif board[x][y] == 'O':
                   board[x][y] = 'X'
        